### PR TITLE
Feature/subscribe event

### PR DESCRIPTION
**Pull request by yazhyvyi**
Subscribe feature, also able to view subscribed users (if logged in)

### CODE REVIEW

File: client/.gitignore [hunk 1]
```diff
@@ -20,3 +20,4 @@ npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
 .eslintrc
+package-lock.json
```
Comment (Merteg) [inline client/.gitignore:23]: I don't think, this file must be here

File: client/src/components/EventPage/index.jsx [hunk 2]
```diff
@@ -52,12 +107,82 @@ class EventPage extends Component {
...
+          })
+        }
+        else {
+          this.setState({
+            is_subbed: false,
+            sub_icon: 'check_circle_outlined',
+            sub_label: 'Subscribe'
+          })
+        }
+        if (data.subscriptions !== null) {
```
Comment (Merteg) [inline client/src/components/EventPage/index.jsx]: if (data.subscriptions) seems better

File: client/src/components/EventPage/index.jsx [hunk 3]
```diff
@@ -52,12 +107,82 @@ class EventPage extends Component {
...
+          this.setState({
+            subs: data.subscriptions
+          })
+        }
+        else {
+          this.setState({
+            subs: null
```
Comment (Merteg) [inline client/src/components/EventPage/index.jsx]: When you define this state, you write `subs: []`, but here `subs: null`, why?

File: client/src/components/EventPage/index.jsx [hunk 4]
```diff
@@ -20,6 +20,11 @@ class EventPage extends Component {
     long: '',
     category: '',
     tags: {},
+    is_subbed: false,
+    sub_icon: 'check_circle_outlined',
+    sub_label: 'Subscribe',
```
Comment (Merteg) [inline client/src/components/EventPage/index.jsx:26]: I'm not sure but may be better to leave this states empty by default, you set this data in componentDidMount method.

File: client/src/components/EventPage/index.jsx [hunk 5]
```diff
@@ -36,6 +41,56 @@ class EventPage extends Component {
...
   }
 
+  SubscribeEvent = (eve) => {
+    eve.preventDefault()
+
+    let data;
+    if (sessionStorage['User-id'] !== undefined) {
+      if (this.state.is_subbed === true) {
+        data = {
+          'action': false,
```
Comment (Merteg) [inline client/src/components/EventPage/index.jsx]: `action`, what it means? I'm don't understand.

File: client/src/components/Home.jsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import React, { Component } from 'react';
```
Comment (Merteg) [inline client/src/components/Home.jsx]: Why did you raise this file from the dead?

**Review by Merteg — CHANGES_REQUESTED** [review]
Hey, this doesn't work, I mean absolutely. You don't have subscribe view.

File: client/src/components/EventPage/index.jsx [hunk 7]
```diff
@@ -36,6 +41,56 @@ class EventPage extends Component {
     }
   }
 
+  SubscribeEvent = (eve) => {
+    eve.preventDefault()
+
+    let data;
+    if (sessionStorage['User-id'] !== undefined) {
+      if (this.state.is_subbed === true) {
```
Comment (serge30) [inline client/src/components/EventPage/index.jsx]: Why? What is the reasons  to compare bool variable to `true` or `false`?

**Comment by yazhyvyi** [comment]
Update

File: server/server/views/subscribe.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
+from cornice.resource import Service
+from pyramid.security import Allow, Everyone
+from ..models import model_to_dict
+from ..models.event import Event
+from ..models.subscribe import Subscribe
+from ..models import model_to_dict
+from ..models.event import Event
+from ..models.subscribe import Subscribe
```
Comment (Merteg) [inline server/server/views/subscribe.py]: Duplicate imports

File: server/server/views/subscribe.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class SubscribeView(object):
```
Comment (Merteg) [inline server/server/views/subscribe.py]: This seems like resource, can you rename it to SubscribeFactory?

File: server/server/views/subscribe.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+     def __init__(self, request):
+         self.request = request
+
+     def __acl__(self):
+         return [(Allow, Everyone, 'get')]
+
+subscribe = Service(name='subscribe', path='/subscribe/{subscribe_user_id}', factory=SubscribeView, cors_origins=('http://localhost:3000',))
```
Comment (Merteg) [inline server/server/views/subscribe.py]: PEP8

**Review by Merteg — CHANGES_REQUESTED** [review]

**Comment by yazhyvyi** [comment]
Update

**Comment by yazhyvyi** [comment]
Update

**Review by Merteg — APPROVED** [review]

**Review by RayMet — APPROVED** [review]

File: client/src/components/EventPage/index.jsx [hunk 11]
```diff
@@ -53,12 +108,82 @@ class EventPage extends Component {
...
+            is_subbed: true,
+            sub_icon: 'check_circle',
+            sub_label: 'Unsubscribe'
+          })
+        }
+        else {
+          this.setState({
+            is_subbed: false,
+            sub_icon: 'check_circle_outlined',
+            sub_label: 'Subscribe'
```
Comment (Merteg) [inline client/src/components/EventPage/index.jsx]: You have this states by default, so you don't need to reset it again.

File: client/src/components/EventPage/index.jsx [hunk 12]
```diff
@@ -53,12 +108,82 @@ class EventPage extends Component {
...
+          })
+        }
+        if (data.subscriptions) {
+          this.setState({
+            subs: data.subscriptions
+          })
+        }
+        else {
+          this.setState({
+            subs: []
```
Comment (Merteg) [inline client/src/components/EventPage/index.jsx]: Also, `subs: []` by default

File: client/src/components/EventPage/index.jsx [hunk 13]
```diff
@@ -53,12 +108,82 @@ class EventPage extends Component {
...
       };
     });
   }
 
+  renderSubscribedUsers() {
+    if (this.state.any_subs) {
+      return (
+        <React.Fragment>
+        {this.state.subs.map((element) => {
+          if (element.avatar === null) {
```
Comment (Merteg) [inline client/src/components/EventPage/index.jsx]: You duplicate much code here just for an avatar. Better to set it to some variable and just call it later.
Like this: `let avatar = element.avatar === null ? "/person.jpg" : element.avatar` and later ...`src={avatar}`...

File: server/server/models/subscribe.py [hunk 14]
```diff
@@ -17,3 +17,27 @@ class Subscribe(Base):
 
     users = relationship("User", foreign_keys=(user_id,))
     events = relationship("Event", foreign_keys=(event_id,))
+
+    @classmethod
+    def get_subscription(cls, request, **kwargs):
+        subscription = request.dbsession.query(cls).filter_by(**kwargs).\
+                       one_or_none()
+        return subscription
```
Comment (Merteg) [inline server/server/models/subscribe.py]: You don't need to create new variable, just `return request.dbsession.query(cls).filter_by(**kwargs).one_or_none()`

File: server/server/views/event.py [hunk 15]
```diff
@@ -128,8 +130,24 @@ def get(self):
...
-        response['status_str'] = self.event_status
-
+            response['status_str'] = self.event_status
+            subscriptions = Subscribe.get_all_subs(request,
+                            event_id=event_obj.id)
+            subscription = Subscribe.get_subscription(request,
+                           user_id=request.user.id,
+                           event_id=event_obj.id)
+            if subscriptions is not None:
+                id_array = [obj.user_id for obj in subscriptions]
```
Comment (Merteg) [inline server/server/views/event.py]: You can get user objects by the relationships, without id_array.

**Review by Merteg — CHANGES_REQUESTED** [review]
Sorry, I was hurry when approving PR, now I see some problems. Also, check PEP8 in your py files.

**Comment by yazhyvyi** [comment]
Update

**Review by Merteg — APPROVED** [review]
Goooood, I think

**Pull request merged**